Swiftui shape (#2930)
Codecademy/docs: content/swiftui/concepts/shapes/shapes.md

---
Title: "Shape"
Description: "The Shape protocol is used for implementing built-in and custom shapes in a SwiftUI view."
Subjects:
  - "Mobile Development"
  - "iOS"
Tags:
  - "SwiftUI"
  - "SwiftUI Shapes"
CatalogContent:
  - "learn-swift"
  - "paths/build-ios-apps-with-swiftui"
---

The **`Shape`** protocol is used to create and implement built-in, and custom shapes.

## Instantiating a Shape

The protocol comes with some shapes included such as capsule, circle, ellipse, and rectangle.

```swift
var body: some View {
    VStack {
        Capsule()
        Circle()
        Ellipse()
        Rectangle()
    }
}
```

## Creating Custom Shapes

Shapes that are not built-in can be defined using path operations and should conform to the `Shape` protocol.

```swift
struct someShape: Shape {
    func path(in rect: CGRect) -> Path {
      // Define the shape using path operations
    }
}
```

> <<<<<<< HEAD

Reviewer: ```suggestion

```

Reviewer: Delete

- The function should return a `Path` object that describes the shape contained inside a rectangular frame of reference.

## Example: Custom Triangle Shape

> Since there is no Triangle shape built-in, a custom one can be made using path operations.
> =======

Reviewer: ```suggestion

Since there is no Triangle shape built-in, a custom one can be made using path operations.
```

Reviewer: Delete "======"

- The function should return a `Path` object that describes the shape contained inside a rectangular frame of reference.

## Example: Custom Triangle Shape

Since there is no triangle shape built-in, a custom one can be made using path operations.

> >>>>>>> a7f91915addc7e1fd9a5ae1bcb3a6d8d4db83343

Reviewer: ```suggestion
```

Reviewer: Delete

```swift
import SwiftUI

struct Triangle: Shape {
    func path(in rect: CGRect) -> Path {
        Path { path in
            path.move(to: CGPoint(x: rect.midX, y: rect.minY))
            path.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY))
            path.addLine(to: CGPoint(x: rect.minX, y: rect.maxY))
            path.addLine(to: CGPoint(x: rect.midX, y: rect.minY))
        }
    }
}
```

- By conforming to the `Shape` protocol, a custom shape inherits all the capabilities of a built-in shape.

> <<<<<<< HEAD
> <br>

Reviewer: ```suggestion
```

Reviewer: Delete

### Displaying the Custom Shape
To display the custom shape, it can be called in a view just like any built-in shape.

> =======
> 

Reviewer: ```suggestion
```

Reviewer: Delete

### Displaying the Custom Shape

To display the custom shape, it can be called in a view just like any built-in shape.

> >>>>>>> a7f91915addc7e1fd9a5ae1bcb3a6d8d4db83343

Reviewer: ```suggestion
```

Reviewer: Delete

```swift
var body: some View {
    ZStack {
        // Sets the color of the background
        Color.blue.ignoresSafeArea()
            
        Triangle()
            .foregroundColor(.yellow)
            .frame(width: 180, height: 180)
    }
}
```

> <<<<<<< HEAD
> =======

Reviewer: ```suggestion
```

Reviewer: Delete

This code results in the following output:

> 
> 
> This code results in the following output:
> 
> >>>>>>> a7f91915addc7e1fd9a5ae1bcb3a6d8d4db83343

Reviewer: ```suggestion

```

Reviewer: Delete

---
Title: "Shape"

> Description: "Methods for implementing built-in and custom shapes in a SwiftUI view."

Reviewer: ```suggestion
Description: "The Shape protocol is used for implementing built-in and custom shapes in a SwiftUI view."
```
While we generally omit the title in terms, we generally don't in concepts.

Subjects:
  - "Mobile Development"
  - "iOS"
Tags:
  - "SwiftUI"
  - "SwiftUI Shapes"
CatalogContent:
  - "learn-swift"
  - "paths/build-ios-apps-with-swiftui"
---

The **`Shape`** protocol is used to create and implement built-in, and custom shapes.

## Instantiating a Shape

The protocol comes with some shapes included such as capsule, circle, ellipse, and rectangle.

```swift
var body: some View {
    VStack {
        Capsule()
        Circle()
        Ellipse()
        Rectangle()
    }
}
```

> <br>

Reviewer: ```suggestion
```
Omit HTML. A simple blank line will suffice.

## Creating Custom Shapes

Shapes that are not built-in can be defined using path operations and should conform to the `Shape` protocol.

```swift
struct someShape: Shape {
    func path(in rect: CGRect) -> Path {
      // Define the shape using path operations
    }
}
```

> - The function should return a `Path` object that describes the shape contained inside a rectangular frame of reference.

Reviewer: ```suggestion

- The function should return a `Path` object that describes the shape contained inside a rectangular frame of reference.
```

> ## Example: Custom Triangle Shape

Reviewer: ```suggestion
## Example: Custom Triangle Shape

```

> Since there is no Triangle shape built-in, a custom one can be made using path operations.

Reviewer: ```suggestion
Since there is no triangle shape built-in, a custom one can be made using path operations.
```

```swift
import SwiftUI

struct Triangle: Shape {
    func path(in rect: CGRect) -> Path {
        Path { path in
            path.move(to: CGPoint(x: rect.midX, y: rect.minY))
            path.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY))
            path.addLine(to: CGPoint(x: rect.minX, y: rect.maxY))
            path.addLine(to: CGPoint(x: rect.midX, y: rect.minY))
        }
    }
}
```

- By conforming to the `Shape` protocol, a custom shape inherits all the capabilities of a built-in shape.

> <br>

Reviewer: ```suggestion
```

> ### Displaying the Custom Shape

Reviewer: ```suggestion
### Displaying the Custom Shape

```

To display the custom shape, it can be called in a view just like any built-in shape.

```swift
var body: some View {
    ZStack {
        // Sets the color of the background
        Color.blue.ignoresSafeArea()
            
        Triangle()
            .foregroundColor(.yellow)
            .frame(width: 180, height: 180)
    }
}
```

> ![SwiftUI Shapes](https://raw.githubusercontent.com/Codecademy/docs/main/media/swiftui-shapes-triangle.jpeg The code displays a yellow triangle on a blue background)

Reviewer: ```suggestion

This code results in the following output:

![SwiftUI Shapes](https://raw.githubusercontent.com/Codecademy/docs/main/media/swiftui-shapes-triangle.jpeg The code displays a yellow triangle on a blue background)
```
